dsp/src/lockin.rs

```diff
@@ -0,0 +1,534 @@
+                let (sine, cosine) = libm::sincosf(phase);
+                let sample = *sample as f32;
+                *i = sine * sample;
+                *q = cosine * sample;
+            });
+
+        Some((in_phase, quadrature))
+    }
+
+    /// Filter the in-phase and quadrature signals using the supplied
+    /// biquad IIR. The signal arrays are modified in place.
+    ///
+    /// # Arguments
+    ///
+    /// * `in_phase` - In-phase signal.
+    /// * `quadrature` - Quadrature signal.
+    pub fn filter(&mut self, in_phase: &mut [f32], quadrature: &mut [f32]) {
+        in_phase
+            .iter_mut()
+            .zip(quadrature.iter_mut())
+            .for_each(|(i, q)| {
+                *i = self.iir[0].update(&mut self.iirstate[0], *i);
+                *q = self.iir[1].update(&mut self.iirstate[1], *q);
+            });
+    }
+}
+
+/// Decimate the in-phase and quadrature signals to
+/// `DECIMATED_BUFFER_SIZE`. The ratio of `ADC_SAMPLE_BUFFER_SIZE` to
+/// `DECIMATED_BUFFER_SIZE` must be a power of 2.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
+///
+/// # Returns
+///
+/// The decimated in-phase and quadrature signals.
+pub fn decimate(
+    in_phase: [f32; ADC_SAMPLE_BUFFER_SIZE],
+    quadrature: [f32; ADC_SAMPLE_BUFFER_SIZE],
+) -> ([f32; DECIMATED_BUFFER_SIZE], [f32; DECIMATED_BUFFER_SIZE]) {
+    let n_k = ADC_SAMPLE_BUFFER_SIZE / DECIMATED_BUFFER_SIZE;
+    debug_assert!(
+        ADC_SAMPLE_BUFFER_SIZE == DECIMATED_BUFFER_SIZE || n_k % 2 == 0
+    );
+
+    let mut in_phase_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+    let mut quadrature_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+
+    in_phase_decimated
+        .iter_mut()
+        .zip(quadrature_decimated.iter_mut())
+        .zip(in_phase.iter().step_by(n_k))
+        .zip(quadrature.iter().step_by(n_k))
+        .for_each(|(((i_decimated, q_decimated), i_original), q_original)| {
+            *i_decimated = *i_original;
+            *q_decimated = *q_original;
+        });
+
+    (in_phase_decimated, quadrature_decimated)
+}
+
+/// Compute the magnitude and phase from the in-phase and quadrature
+/// signals. The in-phase and quadrature arrays are modified in place.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
+pub fn magnitude_phase(in_phase: &mut [f32], quadrature: &mut [f32]) {
+    in_phase
+        .iter_mut()
+        .zip(quadrature.iter_mut())
+        .for_each(|(i, q)| {
+            let new_i = libm::sqrtf([*i, *q].iter().map(|i| i * i).sum());
+            let new_q = libm::atan2f(*q, *i);
+            *i = new_i;
+            *q = new_q;
+        });
+}
+
+/// Treat the 2-element array as a FIFO. This allows new elements to
+/// be pushed into the array, existing elements to shift back in the
+/// array, and the last element to fall off the array.
+trait Fifo2<T> {
```

Comment:
Move this to a point before it's used - I was very confused by what the `push()` was doing on the array at first, but seeing this makes it all clear.